nuxeo-core/nuxeo-core-api/src/main/java/org/nuxeo/ecm/core/api/CoreSession.java

```diff
@@ -1148,13 +1162,123 @@ DocumentModel restoreToVersion(DocumentRef docRef, DocumentRef versionRef, boole
     // LDAP/database.)
     List<String> getAvailableSecurityPermissions();
 
+    /**
+     * Turns the document into a record.
+     * <p>
+     * A record is a document with specific capabilities related to mandatory retention until a given date, and legal
+     * holds. In addition, its main blob receives special treatment from the document blob manager to make sure it's
+     * never shared with another blob at the storage level, and is deleted as soon as the record is deleted.
+     * <p>
+     * If the document is already a record, this method has no effect.
+     * <p>
+     * The permission {@value org.nuxeo.ecm.core.api.security.SecurityConstants#MAKE_RECORD} is required.
+     *
+     * @param docRef the document
+     * @see #isRecord
+     * @since 11.1
+     */
+    void makeRecord(DocumentRef docRef);
+
+    /**
+     * Checks if the document is a record.
+     *
+     * @param docRef the document
+     * @return {@code true} if the document is a record, {@code false} otherwise
+     * @see #makeRecord
+     * @since 11.1
+     */
+    boolean isRecord(DocumentRef docRef);
+
+    /**
+     * The special date that corresponds to a retention date in the indeterminate future.
+     *
+     * @see #setRetainUntil
+     * @since 11.1
+     */
+    Calendar RETAIN_UNTIL_INDETERMINATE = GregorianCalendar.from(ZonedDateTime.parse("9999-01-01T00:00Z"));
+
+    /**
+     * Sets a retention date to the given value.
+     * <p>
+     * If no previous retention date was set, or if the previous retention date was
+     * {@linkplain #RETAIN_UNTIL_INDETERMINATE indeterminate}, or if the previous retention date was <em>before</em>
+     * the given value, then the retention date is set to the given value.
+     * <p>
+     * If the previous retention date was <em>after</em> the given value (that is, if trying to reduce the retention
+     * time), an exception is thrown.
+     * <p>
+     * If the given value is {@code null} and the previous retention date is in the past (it has already expired), then
+     * the retention date is set to {@code null}.
+     * <p>
+     * The permission {@value org.nuxeo.ecm.core.api.security.SecurityConstants#SET_RETENTION} is required.
+     *
+     * @param docRef the document (a record)
+     * @param retainUntil the new retention date
+     * @param comment an optional comment passed to the associated events
+     * @throws PropertyException if trying to reduce the retention time, or if the document is not a record
+     * @see #getRetainUntil
+     * @see #RETAIN_UNTIL_INDETERMINATE
+     * @since 11.1
+     */
+    void setRetainUntil(DocumentRef docRef, Calendar retainUntil, String comment) throws PropertyException;
```

Comment:
i am wondering is there a specific reason to use `Calendar` type instead of `Instant`

Reply:
Historically the `CoreSession` API has used `Calendar` for dates (like we do for DocumentModel properties), and for consistency it's better at this point to not introduce `java.time` here.